### ISSUE TITLE
LSTM accuracy always 1.000, and the output is not as expected

### ISSUE THREAD

**Issue by sjhddh** [fchollet/keras#274]
Hello,
I am quite new to Keras, I am wondering how keras could extend my pre-processed inputs.

My case is similar to a "sentiment analysis". what I have is large english corpus labeled with their "sentiments", either 0(negative) or 1(positive).
Firstly, I tried to use word2vec to give each labeled sentences a fixed dimension of vector representations.

Hence, my *x_train* is N samples, each with dimension 100. the shape will be (nb_samples, sample_dimension) ==> (5000, 100).
and I also scaled them into between 0 and 1. 
Then, of course, the *y_train* is just one-dimension labels, 0 or 1. shape (nb_labels, label_dimension) ==> (5000, 1)

I hope this input format is "well structured", then I try to write with Keras for an LSTM classification model:
```
def fit(x_train, x_test, y_train, y_test):
    # build model
    model = Sequential()
    model.add(Embedding(100, 256))
    model.add(LSTM(256, 128, activation='tanh', inner_activation='hard_sigmoid', return_sequences=False))
    model.add(Dropout(0.5))
    model.add(Dense(128, 1))
    model.add(Activation('tanh'))
    model.compile(loss='binary_crossentropy', optimizer='rmsprop')

    # train model
    model.fit(x_train, y_train, batch_size=30, nb_epoch=10, show_accuracy=True)
    
    # print test results
    print 'run on test set:'
    score = model.evaluate(x_test, y_test, batch_size=30, show_accuracy=True)
    print 'LSTM test accuracy:', score[1]
    print model.predict(x_test, batch_size=30)
    print y_test
```

there occurs a few puzzles:
Firstly, I actually feed pre-processed vectors into the model, so should I remove the **Embedding** layer at the beginning? However, if do so, LSTM layers' input will be wrong. My sample is 2D, seems like LSTM need a 3D input. what can I do? or is it necessary to keep the **Embedding** layer?

Secondly, if I just run this code, I will receive training outputs like this:
```
Epoch 0
671/671 [==============================] - 7s - loss: 1.7508 - acc.: 1.0000     
Epoch 1
671/671 [==============================] - 8s - loss: 0.7121 - acc.: 1.0000     
Epoch 2
671/671 [==============================] - 8s - loss: 0.7165 - acc.: 1.0000     
Epoch 3
671/671 [==============================] - 8s - loss: 0.7133 - acc.: 1.0000     
Epoch 4
671/671 [==============================] - 8s - loss: 0.7077 - acc.: 1.0000     
Epoch 5
671/671 [==============================] - 8s - loss: 0.7175 - acc.: 1.0000     
Epoch 6
671/671 [==============================] - 8s - loss: 0.6922 - acc.: 1.0000     
Epoch 7
671/671 [==============================] - 8s - loss: 0.7076 - acc.: 1.0000     
Epoch 8
671/671 [==============================] - 7s - loss: 0.6990 - acc.: 1.0000     
Epoch 9
671/671 [==============================] - 8s - loss: 0.7036 - acc.: 1.0000     
run on test set:
119/119 [==============================] - 0s - loss: 0.6914 - acc.: 1.0000     
LSTM test accuracy: 1.0
119/119 [==============================] - 0s
```
The **accuracy** is always 1.000, and when I print out the predictions, seems like the model never predicts a right classification, they are all numbers between 0 and 1
```
[[ 0.53357949]
 [ 0.52476386]
 [ 0.52066977]
 [ 0.55957317]
 [ 0.50465528]
 [ 0.52868442]
....
```
while, instead, the labels should be:
```
[ 0.  1.  1.  1.  0.  1.  1.  1.  1.  1.  0.  1.  1.  0.  1.  1.  0.  0.
  0.  0.  1.  0.  1.  0.  1.  1.  1.  0.  0.  1.  1.  0.  1.  0.  0.  1.
  0.  1.  0.  0.  1.  0.  0.  1.  0.  1.  0.  1.  0.  1.  1.  0.  1.  0.
  0.  0.  0.  0.  0.  1.  1.  1.  0.  1.  1.  0.  1.  1.  0.  0.  1.  1.
  0.  1.  1.  1.  0.  1.  0.  0.  0.  0.  1.  0.  0.  1.  1.  1.  0.  1.
  1.  1.  1.  0.  0.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  0.
  0.  0.  0.  0.  0.  1.  1.  1.  1.  0.  0.]
```
I tried to change several things, like activation = 'sigmoid', or optimizer='sgd', it still gives some unacceptable results.

I believe I have made some huge mistakes here, but I really cannot figure it out. 

Wish to get some help, thx

**Comment by donglixp** [comment]
@sjhddh The input of LSTM should be sequence data. In your case, you should input the text sequence into LSTM directly rather than input a fixed vector. If you want to use the 100-dim vectors as features, you can try MLP. The example file examples/imbd_lstm.py is a good start point.

**Comment by sjhddh** [comment]
@donglixp Thx. I actually really want to use LSTM for this task. I tried MLP, it performs not so well.

**Comment by kotrfa** [comment]
I would like to see how to use vector to LSTM as an input too. Additionally, it would be great to see how to also produce the vector (basically LSTM with N input and output nodes). Thanks

**Comment by simonhughes22** [comment]
Here's how you would use an LSTM on an input of (num_samples,MAX_LEN) where each row is a set of id's for each of your words, starting at 1 (as 0's are used to pad). You need to pad the input vectors \ lists to a constant length using the padding function. I pad to the length of the longest sentence, MAX_LEN. See padding call below.

```python
MAX_LEN = maxlen
X_train = sequence.pad_sequences(X_train, maxlen=MAX_LEN) #30 seems good
X_test  = sequence.pad_sequences(X_test,  maxlen=MAX_LEN)

print('X_train shape:', X_train.shape)
print('X_test shape:', X_test.shape)

embedding_size = 64

print('Build model...')
model = Sequential()
model.add(Embedding(max_features, embedding_size))
model.add(LSTM(embedding_size, 64)) # try using a GRU instead, for fun
model.add(Dense(64, 1))
model.add(Activation('sigmoid'))

# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy', optimizer='adam', class_mode="binary")
```
This will then train the vectors for you. You are correct in that the LSTM or GRU expect 3D inputs, where the 2nd dimension is for the number of words \ tokens, and 3rd dimension represents the vector size per word. If you already have vector representations, you can use the Reshape layer to reformat your input as such:

```python
model = Sequential()
model.add(Reshape((1, vector_len)))
model.add(LSTM(embedding_size, 64)) # try using a GRU instead, for fun
model.add(Dense(64, 1))
model.add(Activation('sigmoid'))
```

Alternatively, you can simply feed in a list of lists of vectors as such:

[v1,v2,v3]

where v1 = [[0, 0.2, 0.5, 0.3, 0.5]] so a nested list of s single list, if that makes sense, which is essentially what reshape is doing above. I'd let keras train the vectors for you to the task at hand personally.

One interesting take on this is doing this instead with characters instead of words. The model structure would be identical which a much larger number of columns, but as you can have much smaller vectors for characters than words that works out ok. That way your model doesn't grow with the size of your vocabulary.

I'd also recommend using the adam optimizer and either a GRU or the new LZS1 layer in place of an LSTM, I've found they work better on my dataset.

**Comment by simonhughes22** [comment]
Note for the above, max_feautures is the number of words. The reshape above is needed as you already have a vector per sentence. If you have a vector per word, you'd feed that directly to the LSTM, making sure you pad each row to a constant number of columns first using vectors full of zeros for the missing words. I don't think keras will do this for you, so you'll have to write that yourself. Also, pad from the left, as keras uses the last token's output to predict the sentence label.

You can also try using a convolutional neural network for this task, also trained on padded sentences with an embedding layer.

**Comment by simonhughes22** [comment]
Finally ( :) ), use model.predict_classes to get 0-1 predictions, the above gives you probabilities. Or simply threshold the outputs at 0.5 so [1 if pre >= 0 else 0 for pred in predictions  ]

The accuracy is probably 1.0 as you likely have a very rare class. The loss isn't however, and that's often more telling. Accuracy is a very poor metric when you have imbalanced classes, I'd suggest computing an F1 score on your test set instead. sklearn has methods for doing this that you can use here.

**Comment by GiterLI** [comment]
@sjhddh I would like to feed existing vector (without learning the vector) into lstm too. Have you solved the problem?

**Comment by sjhddh** [comment]
@GiterLI no, I havent find a better solution with Keras.
Yeah, I am thinking about a solution for this, and I think it could be important for many others who wanna "play" with different models. I got an optional solution finally, but still not so "pure". I am still trying to figure something I can do with Keras.

**Comment by viksit** [comment]
@sjhddh @simonhughes22 any pointers on the last question - utilizing existing embeddings (aka glove/word2vec) into an LSTM without learning embeddings via keras itself?

**Comment by simonhughes22** [comment]
Yeah just feed in as features, then use the same model as before without the embedding layer:

model = Sequential()
model.add(GRU(embedding_size, hidden_size))
model.add(Dense(hidden_size, 1))
model.add(Activation('sigmoid'))

The input will need to be 3D, (num_rows, embedding_size, sequence_length).

What would probably work better is seeding the embeddings with pre-trained vectors and fine-tuning them. For that you would need to inititalize the embeddings with the weights of the vectors. I haven't done that but it should be easy.

**Comment by xisnu** [comment]
@simonhughes22 Thank you sir for your response in this topic. I am trying to use LSTM for a simple classification problem with IRIS data set. My training input is an array [150,4] 150 samples each with 4 features and my target is [150,3], as number of classes is 3. I am using the same set for testing. So I have created the following network
```
model = Sequential()
model.add(Embedding(4,64))
model.add(LSTM(64,activation='tanh', inner_activation='hard_sigmoid'))#Create Input Layer
model.add(Dense(3, init='uniform'))#Create output layer
model.add(Activation('softmax')) 
```
the network is successfully created. Now I am training with the following parameters.
```
model.fit(inputs, targets,nb_epoch=100,batch_size=10,shuffle=True,verbose=1,show_accuracy=True)
```
**inputs** is the [150,4] array and **targets** is the [150,3] array as said previously. The class labels are binary i.e 1,0,0 represents **class A**, 0,1,0 represents **class B** and so on. Now I have used SGD and RMSProp for experiments. But the system is always identifying a particular class. So the accuracy is always around 33.33%. Moreover all the 3 outputs for each of the 150 samples are producing same results. i.e I am getting 150 rows of same data when testing.

What am I doing wrong. May be I am misunderstanding the use of LSTM in Keras completely. I have experimented with the same data in MLP. It gave excellent result. I am just curious in LSTM.

Any suggestion is highly appreciated. Thank you for your time.

**Comment by marchss** [comment]
Any solution for this issue yet?

**Comment by farmeh** [comment]
@simonhughes22 : you had said "Also, pad from the left, as keras uses the last token's output to predict the sentence label.". Can you please explain what do you mean? 
I have written my own function for padding, which add zeros to the end, 
and I have the same problem:
acc is always == 1.0 (negative class is the dominant class), 
and optimization of loss does not converge. 

thanks in advance.

**Comment by earonesty** [comment]
add zeroes to the front, instead

**Comment by fredzilla** [comment]
I can confirm that this doesn't help. In fact, it seems that this error occurs in the IMDb example packaged with Keras.

The IMDb example packaged with Keras doesn't seem to work with the latest version, so I had to change a few lines in order to get it to work. Attached is the code which also exhibits the accuracy = 1 for all cases bug mentioned by @sjhddh.
